plugins/witness_trx_plugin/witness_trx_plugin.cpp

```diff
@@ -0,0 +1,243 @@
+            sig_action_data_waiting_on_catch_up.splice(sig_action_data_waiting_on_catch_up.begin(), row.action_data);
+         });
+      }
+
+      try {
+         fc::datastream<fc::cfile> file;
+         file.set_file_path(trx_witness_path);
+         file.open("wb");
+         fc::raw::pack(file, sig_action_data_waiting_on_catch_up);
+      } FC_LOG_AND_DROP();
+   }
+};
+
+witness_trx_plugin::witness_trx_plugin() : my(new witness_trx_plugin_impl()) {}
+
+void witness_trx_plugin::set_program_options(options_description& cli, options_description& cfg) {
+   auto default_priv_key = eosio::chain::private_key_type::regenerate<fc::ecc::private_key_shim>(fc::sha256::hash(std::string("witness account")));
+
+   cfg.add_options()
+         ("witness-account", boost::program_options::value<string>()->default_value(my->witness_account.to_string())->notifier([this](const string& v) {
+                  my->witness_account = eosio::name(v);
+               }), "EOSIO account to submit witness transactions to")
+         ("witness-trx-signature-provider", boost::program_options::value<string>()->default_value(
+               default_priv_key.get_public_key().to_string() + "=KEY:" + default_priv_key.to_string())->notifier([this](const string& v){
+                  const auto& [pubkey, provider] = app().get_plugin<signature_provider_plugin>().signature_provider_for_specification(v);
+                  my->signature_provider = provider;
+               }),
+               app().get_plugin<signature_provider_plugin>().signature_provider_help_text())
+         ;
+
+   cli.add_options()
+         ("witness-trx-delete-unsent", boost::program_options::bool_switch(&my->delete_previous), "Delete unsent witness signature transactions");
+}
+
+void witness_trx_plugin::plugin_initialize(const variables_map& options) {
+   my->trx_witness_path = appbase::app().data_dir() / "trx_witness.bin";
+
+   const auto chain_read_mode = my->chainplug.chain().get_read_mode();
+   EOS_ASSERT(chain_read_mode == chain::db_read_mode::HEAD || chain_read_mode == chain::db_read_mode::SPECULATIVE, chain::plugin_config_exception,
+              "witness_trx_plugin requires head or speculative chain mode");
+}
+
+void witness_trx_plugin::plugin_startup() {
+   app().get_plugin<witness_plugin>().add_on_witness_sig([my=my](const chain::block_state_ptr& bsp, const chain::signature_type& sig) {
+      //if nodeos is shutting down, it is possible witness_trx_plugin_impl will be destroyed immediately after the callback returns before
+      // the post() below is actually dispatched. Prevent returning until certain post() has dispatched.
+      std::promise<void> made_to_other_thread_promise;
+      auto fut = made_to_other_thread_promise.get_future();
+
+      my->ctx.post([my, &bsp, &sig, &made_to_other_thread_promise]() {
+         std::vector<char> witness_action_data = fc::raw::pack(std::make_pair(bsp->header.action_mroot, sig));
+         made_to_other_thread_promise.set_value();
```

Comment:
If nodeos is shutting down, then once we notify the calling thread, the rest of nodeos could run to completion including returning from main, before the rest of this function executes.  We'll be left with a dangling reference to `chainplug` in that case AND the worker thread will join itself.

Reply:
chainplug is easy to refactor out; it's stuff like the `app()` access that would be harder. Starting to like the old `weak_ptr` approach more...

Reply:
Wouldn't the `weak_ptr` approach only affect the outer callback, not this inner one?

Reply:
I think `weak_ptr` may make it easier to deal with something like `app()`; or more generally, ensuring all dependent plugin_impls are destroyed before witness_plugin_impl stop() returns. But yeah the self join() problem remains.